lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +29,194 @@ pub fn handle_completion(
+        linearization: &Vec<LinearizationItem<Types>>,
+        id: usize,
+    ) -> Option<Contract> {
+        linearization.iter().find_map(|item| match &item.meta {
+            Some(MetaValue { contracts, .. }) if item.id == id => {
+                contracts.iter().find_map(|contract| {
+                    if let Types(AbsType::Record(..)) = contract.types {
+                        Some(contract.clone())
+                    } else {
+                        None
+                    }
+                })
+            }
+            _ if item.id == id => match item.kind {
+                TermKind::Declaration(_, _, ValueState::Known(new_id))
+                | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                    find_record_contract(&linearization, new_id)
+                }
+                _ => None,
+            },
+            _ => None,
+        })
+    }
 
+    /// Extract identifiers from a row type which forms a record.
+    /// Ensure that ty is really a row type.
+    fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+        ty.iter_as_rows()
+            .filter_map(|item| match item {
+                RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+                RowIteratorItem::Tail(..) => None,
+            })
+            .collect::<Vec<_>>()
+    }
+
+    let item = linearization.item_at(&locator);
     if item == None {
+        // NOTE: this might result in a case where completion is not working.
+        // This happens when an item cannot be found from the linearization.
         server.reply(Response::new_ok(id, Value::Null));
         return Ok(());
     }
-
     let item = item.unwrap().to_owned();
 
-    let in_scope: Vec<_> = linearization
-        .get_in_scope(&item)
+    Trace::enrich(&id, linearization);
+
+    let trigger = params.context.and_then(|context| context.trigger_character);
+    let in_scope = match trigger {
+        // Record Completion
+        Some(s) if s == "." => {
+            // Get the ID before the .
+            // if the current source is `let var = {a = 10} in var.`
+            // name = "var"
+            let name = text[..start - 1]
+                .chars()
+                .rev()
+                .skip_while(|c| *c == '.') // skip . (if any)
+                .take_while(|c| c.is_ascii_alphabetic())
+                .collect::<String>()
+                .chars()
+                .rev()
+                .collect::<String>();
+
+            let name_id = lin_env
+                .iter()
+                .find(|(ident, _)| ident.label() == name)
+                .map(|(_, id)| *id);
+            if name_id == None {
+                // NOTE: this might result in a case where completion is not working.
+                // This happens when an item cannot be found from the linearization.
+                server.reply(Response::new_ok(id, Value::Null));
+                return Ok(());
+            }
+            let name_id = name_id.unwrap();
```

Comment:
Once again this might be a total noob question, please bear with me :pray: 

If I understand this code correctly, we unsafely extract from the `Some` after handling the `None` case and early-returning.

This might be dangerous in the face of refactoring: e.g. moving the `if` below would throw at runtime.

Can we use pattern matching instead? That way, there is a guarantee of safety "for free":
```rust
match name_id {
  None => {
    server.reply(Response::new_ok(id, Value::Null));
    return Ok(());
  }
  Some(name_id) => {...}
}
```

Reply:
I think the reason it is like that is to avoid the nesting that comes with that pattern matching. But I think we can also have this safety guarantee by using the `?` short-circuiting operator:
```rust
 let item = linearization.item_at(&locator);
 let item = item.ok_or(ResponseError {
        code: ErrorCode::InternalError as i32,
        message: "Couldn't get the linearization item at the cursor position".to_owned(),
        data: None,
    })?;
```

But this way, it doesn't return an `Ok(..)` response

Reply:
So what is the idiom in Rust, or in the Nickel codebase in particular?
I would pick "safe and future proof" over "less indentation level" every day of the year, but I don't want to enforce my opinion on the codebase :smile:

Reply:
Side note: maybe the problem is also that this function is too big, and takes care of too many things? It is currently 156 lines long, which might be an indicator.

Reply:
I tend to agree with @sir4ur0n on this one. It doesn't feel right to do an `unwrap` when we can prove easily at compile time that there is no way of panicking (as a side note, although in this case the compiler may be able to optimize the check away, it also has a potential runtime cost, because `unwrap` need to check again that the value is not `None`. It's probably highly negligible in practice, so it's more of a psychological thing, but still)